### ISSUE TITLE
Add nodes to running cluster

### ISSUE THREAD

**Issue by FxKu** [kubernetes-sigs/kind#452]
**What would you like to be documented**:
How can I add more nodes to a running multi-node cluster? Do I have to use kubeadm for that? Is it even possible with kind?

**Why is this needed**:
I want to test a scenario where a whole cluster is upgraded, but not by upgrading running nodes, but by creating a new worker pool and then switched over to them in rolling update fashion.

**Comment by BenTheElder** [comment]
Experimental upgrade work for testing kubeadm is in https://github.com/kubernetes/kubeadm/tree/master/kinder

Currently we do not support adding nodes or upgrading in kind. We might in the future.

**Comment by FxKu** [comment]
Thanks for the quick info and pointer to kinder. Closing for now.

**Comment by abdennour** [comment]
any news about this feature ?

**Comment by BenTheElder** [comment]
no, we've made it faster to create and delete clusters, work like this is on hold while we refactor how nodes are managed entirely

**Comment by akutz** [comment]
Hi @bentheelder,

I was just in the shower trying to devise an e2e test framework for CAPV, with Kind being, of course, a huge part of it. The framework would be one part vC Simulator and one part Kind. However, this falls down if we can’t add/remove nodes to a running Kind cluster. There’s still some basic testing we can do by preallocating the max number of nodes we know we’ll need, but this obviously remove the ability to test adding/removing nodes in a true e2e fashion.

Basically this is a big priority for CAPV, and anything I can do to help prioritize this issue, please let me know.

**Comment by qoobic** [comment]
@BenTheElder Is there any update on this please? We're trying to simulate auto-scaling behaviour, my plan was to manually add nodes to the KIND cluster but it seems that isn't possible. Is it possible to add a label to a node while it is running, although I don't know if that would even work if K8S caches any of the node configuration?

**Comment by BenTheElder** [comment]
No, there is not. We don't have the bandwidth for this and it's not really a key part of what we're trying to provide right now.
You can scale a local cluster with https://github.com/kubernetes-sigs/cluster-api-provider-docker which is based on KIND.

You can usually simulate scaling behavior with taints unless you have a controller monitoring the node object. In which case you can simulate it with fake nodes (see kubemark).

**Comment by qoobic** [comment]
Thanks @BenTheElder. I have accomplished what I needed using labels as I was able to edit the node configuration via `kubectl`. But the info you provided will be helpful, so thanks for that.

**Comment by migueleliasweb** [comment]
Being able to add nodes to a running cluster is also useful for testing operators that listen to new nodes in a cluster.

**Comment by aojea** [comment]
This is something that people can develop using the kind API , some examples on how to use.it here https://github.com/aojea/kind-networking-plugins

As Ben says, that is a very complex feature to maintain

**Comment by migueleliasweb** [comment]
By "kind API" do you mean cAPI as in https://cluster-api.sigs.k8s.io/ ?

**Comment by aojea** [comment]
I mean someone can create a golang program that implements this feature, the linked repo shows some examples on how someone can create these extensions

**Comment by BenTheElder** [comment]
Cluster API does collaborate with us and has a KIND-based "docker provider" built in that can do more dynamic clusters. It used to be at the repo linked above but you can find the current implementation in the main cluster API project I think.

I don't see where "kind API" was mentioned here but that would refer to the go packages Antonio mentioned.

**Comment by migueleliasweb** [comment]
Hi @BenTheElder first of all, thanks for creating KinD. It's a beautiful thing :heart_eyes: .

With that out of the way.

> Cluster API does collaborate with us and has a KIND-based "docker provider" built in that can do more dynamic clusters.

I've read this various times in different issues but I'm not gonna lie I could never understand what this really mean. I know cAPI **does** have a bootstrap/management provider that uses Kind but I'm not sure if this is exactly what the original author was talking about. I think it was more about the `workload` cluster instead of the `management` one.

I tried to find a way to target Kind directly from `clusterctl` but that doesn't seem supported for workload clusters, as far as I can tell.

This is from the cAPI website:

![Screenshot from 2022-11-21 09-01-02](https://user-images.githubusercontent.com/1608212/202928638-5ee92565-fbf0-4f78-876f-389743f18ad0.png)

**Comment by BenTheElder** [comment]
There's a cluster api provider called something like "docker" which is based on kind, distinct from the bootstrap provider. You can use the docker provider to create a kind-like cluster that can be dynamically reconfigured with cluster API. It used to be in the repo linked above, it's in the main cluster API project now I think

**Comment by BenTheElder** [comment]
https://cluster-api.sigs.k8s.io/user/quick-start.html "Docker" provider option used to be https://github.com/kubernetes-sigs/cluster-api-provider-docker

**Comment by lobuhi** [comment]
Solved https://github.com/lobuhi/kindscaler
